### PR TITLE
Architectural refactoring

### PR DESCRIPTION
**Pull request by flit**
Fairly significant changes to re-architect pyOCD's target and Cortex-M related code.

- Created new pyOCD.coresight submodule. It has DebugPort and AccessPort classes, as well as classes for CortexM, FPB, DWT, breakpoints, and ROM table parsing.
- CortexM class moved to pyOCD.coresight and simplified by using other coresight component classes such as AHB_AP and the breakpoints classes.
- New CoreSightTarget subclass of Target that owns the DebugPort, AccessPort, and CortexM classes. It supports multiple cores. The existing target classes are now subclasses of CoreSightTarget rather than CortexM.
- Factored breakpoints code that was in CortexM into a BreakpointsManager class and related classes.
- CoreSight ROM tables are parsed during init, although they are not yet used to construct objects.
- Additional CPU states are supported: reset, lockup, and sleep.
- Debug resources are disabled and debug is powered down upon disconnect.
- Memory reads are filtered to hide software breakpoints.

This changeset also includes some miscellaneous changes, as well:
- pyocd-tool can execute arbitrary Python expressions via a "$<python code>" syntax.
- Added "core" command to pyocd-tool to select the current core.
- SVD files are associated with targets and the registers can be accessed via pyocd-tool.
- hidapi USB backend was improved to fix issues with multiple connected boards and when another process already has a board open.

### CODE REVIEW

**Comment by 0xc0170** [comment]
Chris, thanks for pushing this upstream :+1: . I'll have a look this week for all these changes in here

**Comment by tgarc** [comment]
@flit are you one of the core contributors or are you talking with the core contributors outside of github? I'm just assuming that you have been because otherwise merging 45+ commits that add all kinds of different functionality seems kind of insane. I really am trying to say this to be constructive, I just think PRs should be fairly succinct in what they're trying to achieve. The set of changes your proposing is on the level of an entirely new release.

@0xc0170 @c1728p9 Thoughts?

**Comment by 0xc0170** [comment]
> are you one of the core contributors or are you talking with the core contributors outside of github?

Actually both.

>  I'm just assuming that you have been because otherwise merging 45+ commits that add all kinds of different functionality seems kind of insane.

Do you think, Chris we could split this into smaller feature-complete PR? At least we attempt to reduce the number of features here.

**Comment by c1728p9** [comment]
Excellent work on this patch @flit !  This patch should probably be followed by a minor version bump if it can't be easily broken into a lot of smaller patches.  I'll try to give it a thorough first pass review here though.  Thanks again for all the hard work.  Sorry I wasn't able to get to it sooner.  I have been quite busy lately.

File: pyOCD/coresight/ap.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+            return res
+
+        if now:
+            result = readMemCb()
+            return result
+        else:
+            return readMemCb
+
+    # write aligned word ("data" are words)
+    def writeBlock32(self, addr, data):
```
Comment (c1728p9) [inline pyOCD/coresight/ap.py]: this should probably be made private since it can only be used correctly if auto_increment_page_size is known.

File: pyOCD/coresight/ap.py [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+        if (size > 0):
+            self.writeMemory(addr, data[idx], 8)
+            size -= 1
+            addr += 1
+            idx += 1
+
+        return
+
+    ## @brief Write a block of aligned words in memory.
+    def writeBlockMemoryAligned32(self, addr, data):
```
Comment (c1728p9) [inline pyOCD/coresight/ap.py]: I'm not sure if there is much of a need to write blocks of 32 bytes values.  It might be good to make this private/remove it from the public API.  What do you think?

File: pyOCD/coresight/cortex_m.py [hunk 3]
```diff
@@ -335,66 +284,47 @@ def __init__(self, name, bitsize, reg_type, reg_group):
...
         self.has_fpu = False
-        self.part_number = self.__class__.__name__
-        self.dap = Dap(link)
+        self.dp = dp
+        self.ap = ap
+        self.core_number = core_num
+
+        # Set up breakpoints manager.
+        self.fpb = FPB(self.ap)
+        self.dwt = DWT(self.ap)
```
Comment (c1728p9) [inline pyOCD/coresight/cortex_m.py]: In the future will the FPB and DWT be created automatically based on info in the rom table?

File: pyOCD/coresight/cortex_m.py [hunk 4]
```diff
@@ -335,66 +284,47 @@ def __init__(self, name, bitsize, reg_type, reg_group):
...
+
+    def disconnect(self):
+        # Remove breakpoints.
+        self.bp_manager.remove_all_breakpoints()
+
+        # Disable core debug.
+        self.write32(CortexM.DHCSR, CortexM.DBGKEY | 0x0000)
+
+        # Disable other debug blocks.
+        self.write32(CortexM.DEMCR, 0)
```
Comment (c1728p9) [inline pyOCD/coresight/cortex_m.py]: Should this also disable power in the underlying DP?  Or is this managed somewhere else or unnecessary?  Also did you find out if the debug logic can be turned off to reduce current consumption on Kinetis devices?

File: pyOCD/coresight/cortex_m.py [hunk 5]
```diff
@@ -742,11 +496,28 @@ def reset(self, software_reset=None):
...
         if software_reset:
-            self.writeMemory(CortexM.NVIC_AIRCR, CortexM.NVIC_AIRCR_VECTKEY | CortexM.NVIC_AIRCR_SYSRESETREQ)
-            # Without a flush a transfer error can occur
-            self.flush()
+            # Perform the reset.
+            try:
+                self.writeMemory(CortexM.NVIC_AIRCR, CortexM.NVIC_AIRCR_VECTKEY | CortexM.NVIC_AIRCR_SYSRESETREQ)
+                # Without a flush a transfer error can occur
+                self.dp.flush()
+            except DAPAccess.TransferError:
```
Comment (c1728p9) [inline pyOCD/coresight/cortex_m.py]: Why is there a try except block here?  Is reset causing a transfer error?

File: pyOCD/coresight/cortex_m.py [hunk 6]
```diff
@@ -876,11 +665,15 @@ def readCoreRegistersRaw(self, reg_list):
...
             dhcsr_val = dhcsr_cb()
-            assert dhcsr_val & CortexM.S_REGRDY
+#            assert dhcsr_val & CortexM.S_REGRDY
             val = reg_cb()
 
-            # Special handling for registers that are combined
-            # into a single DCRSR number.
+            # If we couldn't read the register, stuff a dummy zero in the return value list.
+            if (dhcsr_val & CortexM.S_REGRDY) == 0:
+                reg_vals.append(0)
```
Comment (c1728p9) [inline pyOCD/coresight/cortex_m.py]: Instead of adding a dummy value here, could an exception be raised?  The calling code could then catch the appropriate exception if they want to be able to recover from this.

File: pyOCD/coresight/cortex_m.py [hunk 7]
```diff
@@ -876,11 +665,15 @@ def readCoreRegistersRaw(self, reg_list):
...
+                continue
```
Comment (c1728p9) [inline pyOCD/coresight/cortex_m.py]: Do some of the register get read correctly and others not?

File: pyOCD/flash/flash_kl28z.py [hunk 8]
```diff
@@ -79,9 +80,41 @@
...
+
+        # Enable FIRC.
+        value = self.target.read32(SCG_FIRCCSR)
+        value |= FIRCEN_MASK
+        self.target.write32(SCG_FIRCCSR, value)
+
+        # Switch system to FIRC, core=48MHz (/1), slow=24MHz (/2).
+        # Flash and the bus are clocked from the slow clock, and its max is 24MHz,
+        # so there is no benefit from raising the core clock further.
+        self.target.write32(SCG_RCCR, (0x3 << SCS_SHIFT) | (1 << DIVSLOW_SHIFT))
```
Comment (c1728p9) [inline pyOCD/flash/flash_kl28z.py]: So does bumping up the clock speed up flash programming a lot?  I have been thinking about trying this on other devices to allow them to flash faster.

Also, would it make sense to move this logic into the flash algo?  That way no target specific configuration would be needed.

File: pyOCD/svd.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+import threading
+from cmsis_svd.parser import SVDParser
```
Comment (c1728p9) [inline pyOCD/svd.py]: The package "cmsis-svd" needs to be added as a dependency so it is installed automatically.

File: pyOCD/target/__init__.py [hunk 10]
```diff
@@ -41,7 +41,7 @@
 import semihost
 
 TARGET = {
-          'cortex_m': cortex_m.CortexM,
+#           'cortex_m': cortex_m.CortexM,
```
Comment (c1728p9) [inline pyOCD/target/__init__.py]: Why was cortex_m removed?

File: pyOCD/coresight/cortex_m.py [hunk 11]
```diff
@@ -335,66 +284,47 @@ def __init__(self, name, bitsize, reg_type, reg_group):
...
+
+        # Set up breakpoints manager.
+        self.fpb = FPB(self.ap)
+        self.dwt = DWT(self.ap)
+        self.sw_bp = SoftwareBreakpointProvider(self)
+        self.bp_manager = BreakpointManager(self)
+        self.bp_manager.add_provider(self.fpb, Target.BREAKPOINT_HW)
+        self.bp_manager.add_provider(self.sw_bp, Target.BREAKPOINT_SW)
 
     def init(self, initial_setup=True, bus_accessible=True):
```
Comment (c1728p9) [inline pyOCD/coresight/cortex_m.py]: The parameters 'initial_setup' and 'bus_accessible' can probably be removed since the init sequence is a lot simpler.

File: pyOCD/target/target_k20d50m.py [hunk 12]
```diff
@@ -30,4 +31,5 @@ class K20D50M(Kinetis):
     def __init__(self, link):
         super(K20D50M, self).__init__(link, self.memoryMap)
         self.mdm_idr = 0x001c0000
+        self._svd_location = SVDFile(vendor="Freescale", filename="MK20D5.xml", is_local=False)
```
Comment (c1728p9) [inline pyOCD/target/target_k20d50m.py]: Do the svd files need to be copied locally to be used?  What can be done with the svd files?  Can they be easily be used from within python, or are they only useful for the pyocd-tool?

File: pyOCD/tools/gdb_server.py [hunk 13]
```diff
@@ -178,12 +178,12 @@ def list_boards(self):
                     'info' : mbed.getInfo(),
                     'board_name' : mbed.getBoardName(),
                     'target' : mbed.getTargetType(),
-                    'vendor_name' : mbed.interface.vendor_name,
-                    'product_name' : mbed.interface.product_name,
+                    'vendor_name' : "", #mbed.interface.vendor_name,
+                    'product_name' : "", #mbed.interface.product_name,
```
Comment (c1728p9) [inline pyOCD/tools/gdb_server.py]: Why were these removed?  pyDAPAccess changes?  Should a todo be added?

File: pyOCD/coresight/cortex_m.py [hunk 14]
```diff
@@ -451,244 +381,69 @@ def checkForFPU(self):
...
-        Shorthand to write a 16-bit halfword.
-        """
-        self.writeMemory(addr, value, 16)
-
-    def write8(self, addr, value):
-        """
-        Shorthand to write a byte.
-        """
-        self.writeMemory(addr, value, 8)
+        self.ap.writeMemory(addr, value, transfer_size)
```
Comment (c1728p9) [inline pyOCD/coresight/cortex_m.py]: Should memory be filtered on writes so breakpoints don't get overwritten?

File: pyOCD/coresight/cortex_m.py (hunk 4 above)
Comment (c1728p9) [inline pyOCD/coresight/cortex_m.py]: Is there any reason the DEMCR is disabled before the DHCSR disabled debugging?  It might be safer to do in the reverse order.

File: pyOCD/tools/flash_tool.py [hunk 15]
```diff
@@ -45,7 +45,10 @@
 
 supported_formats = ['bin', 'hex']
 supported_targets = pyOCD.target.TARGET.keys()
-supported_targets.remove('cortex_m')  # No generic programming
+try:
+    supported_targets.remove('cortex_m')  # No generic programming
+except:
```
Comment (c1728p9) [inline pyOCD/tools/flash_tool.py]: Why is there a try except block here?

File: pyOCD/target/target_kl28z.py [hunk 16]
```diff
@@ -60,12 +74,32 @@ def init(self):
...
+
+            # Add second core. It is held in reset until released by software.
+            self.core1 = CortexM(self.link, self.dp, self.core1_ap, self.memory_map, core_num=1)
+            self.cores[1] = self.core1
+            self.core1.init(bus_accessible=True)
+
+        # Disable ROM vector table remapping.
+        self.write32(RCM_MR, RCM_MR_BOOTROM_MASK)
+
+    def recover_from_fault(self, error):
```
Comment (c1728p9) [inline pyOCD/target/target_kl28z.py]: Why does this target need a custom recovery handler?

**Comment by c1728p9** [comment]
A lot of small comments, but in general the change looks great!  I'll be running some more testing tomorrow, so there will probably be some more feedback coming.

Found a couple of problems while testing.  I'll try to help track them down more tomorrow.
-lpc800, lpc1768 and lpc11u24 fail basic test, cortex test and flash test.
-the gdb test script no longer runs properly and hangs.  It looks like the response to setting a software breakpoint changed, which might be causing this.  I have attached logs from this test before and after your patch.
[gdb_before_and_after.zip](https://github.com/mbedmicro/pyOCD/files/66036/gdb_before_and_after.zip)

**Comment by c1728p9** [comment]
Anyone who is interested feel free to test this patch and provide feedback.  Also please comment if there are any specific concerns you would like to have addressed.  I'll be working with @flit to make this as stable as possible before it is merged.

File: pyOCD/target/target.py [hunk 17]
```diff
@@ -15,11 +15,13 @@
  limitations under the License.
```
Comment (c1728p9) [inline pyOCD/target/target.py]: Any reason MKL28T7_CORE0.svd is checked in but no other svd files are?

File: pyOCD/svd.py (hunk 9 above)
Comment (posborne) [inline pyOCD/svd.py]: Note that the `cmsis_svd` package is currently quite large as it includes a collection of SVD files themselves.  If this a problem, we can probably figure something out.  Just create an issue with upstream: https://github.com/posborne/cmsis-svd (I'm the maintainer).

File: pyOCD/target/target_k20d50m.py (hunk 12 above)
Comment (posborne) [inline pyOCD/target/target_k20d50m.py]: A pretty wide selection of SVD files ship with the `cmsis_svd` package.  New SVDs can be added by sending PRs to the cmsis-svd project: https://github.com/posborne/cmsis-svd/tree/master/data

Comment (0xc0170) [inline pyOCD/target/target_k20d50m.py]: +1 @posborne for answering questions here

**Comment by flit** [comment]
Thanks @c1728p9 for reviewing and all your comments, sorry it took a while to get back to this. Also thanks for testing. I've seen the gdb test failure, but wasn't sure if it was something I broke or the fact I was running on OS X. Not sure why the LPC targets would fail… will look into that.

@tgarc yeah, it's a lot of commits… But a lot of the commits have actually already been merged via other pull requests. In those cases the work was originally done on this branch, then I moved it to another branch and created a PR. I chose not to rebase this branch because there were quite complicated differences to resolve, especially regarding the new pyDAPAccess submodule.

File: pyOCD/coresight/ap.py (hunk 1 above)
Comment (flit) [inline pyOCD/coresight/ap.py]: Agreed, we have the read/writeMemoryAligned32() methods for the public API.

**Comment by c1728p9** [comment]
test this please

**Comment by c1728p9** [comment]
test this please

**Comment by c1728p9** [comment]
test this please

File: pyOCD/target/target.py (hunk 17 above)
Comment (flit) [inline pyOCD/target/target.py]: Just committed a change to remove this file from the branch. Created a PR to update cmsis-svd with the latest Kinetis SVD files, including this one (looks like posborne just merged it :)

**Comment by flit** [comment]
@c1728p9 Just committed some changes based on questions you had about the changes in this PR.

1. The retry loop in CortexM.reset() is still needed for KL28. But I removed the fault recovery handler, which seemed to be overkill and wasn't used in my testing.

2. Stuffing the dummy 0 value if S_REGRDY isn't set in CortexM.readCoreRegistersRaw() seems to have been related to multicore. Some code was trying to read registers when the core was being held in reset. I've reverted this code for the time being and am moving it to my multicore debug branch.

3. Cleaned up initial_setup and bus_accessible params for init() of CortexM and CoreSightTarget.

4. Removed the KL28 SVD file in favor of a PR to cmsis-svd, which has already been merged thanks to posborne.

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
Latest patchset is failing with:
```
Traceback (most recent call last):
  File "automated_test.py", line 111, in <module>
    main()
  File "automated_test.py", line 85, in main
    board_list = MbedBoard.getAllConnectedBoards(close=True, blocking=False)
  File "/home/workspace/pyOCD_hw/pyOCD/board/mbed_board.py", line 163, in getAllConnectedBoards
    new_mbed = MbedBoard(dap_access, target_override, frequency)
  File "/home/workspace/pyOCD_hw/pyOCD/board/mbed_board.py", line 102, in __init__
    super(MbedBoard, self).__init__(target, target, link, frequency)
  File "/home/workspace/pyOCD_hw/pyOCD/board/board.py", line 30, in __init__
    self.target = TARGET[target](self.link)
  File "/home/workspace/pyOCD_hw/pyOCD/target/target_lpc1768.py", line 32, in __init__
    self._svd_location = SVDFile(vendor="NXP", filename="LPC176x5x_v0.2.svd", is_local=False)
NameError: global name 'SVDFile' is not defined
```

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by 0xc0170** [comment]
Only cortex-m this time

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by flit** [comment]
@c1728p9 is this latest failure caused by the code, or is it an issue with the mbed CI server?

**Comment by c1728p9** [comment]
It is likely a problem with CI

**Comment by c1728p9** [comment]
@mbed-bot: TEST

**Comment by c1728p9** [comment]
@mbed-bot: TEST
